services/leader/client.go

```diff
@@ -0,0 +1,245 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package leader
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/leader/campaign"
+	"github.com/m3db/m3cluster/services/leader/election"
+
+	"golang.org/x/net/context"
+)
+
+// Appended to elections with an empty string for electionID to make it easier
+// for user to debug etcd keys.
+const defaultElectionID = "default"
+
+var (
+	// ErrNoLeader is returned when a call to Leader() is made to an election
+	// with no leader. We duplicate this error so the user doesn't have to
+	// import etcd's concurrency package in order to check the cause of the
+	// error.
+	ErrNoLeader = concurrency.ErrElectionNoLeader
+
+	// ErrCampaignInProgress is returned when a call to Campaign() is made while
+	// the caller is either already (a) campaigning or (b) the leader.
+	ErrCampaignInProgress = errors.New("campaign in progress")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.RWMutex
+
+	client      *election.Client
+	opts        services.ElectionOptions
+	cancelFn    context.CancelFunc
+	resignCh    chan struct{}
+	campaigning bool
+	closed      bool
+}
+
+// newClient returns an instance of an client bound to a single election.
+func newClient(cli *clientv3.Client, opts Options, electionID string) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	ttl := opts.ElectionOpts().TTLSecs()
+	pfx := electionPrefix(opts.ServiceID(), electionID)
+	ec, err := election.NewClient(cli, pfx, election.WithSessionOptions(concurrency.WithTTL(ttl)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		client:   ec,
+		opts:     opts.ElectionOpts(),
+		resignCh: make(chan struct{}),
+	}, nil
+}
+
+func (c *client) campaign(opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !c.startCampaign() {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c.Lock()
+	c.cancelFn = cancel
+	c.Unlock()
+
+	proposeVal := opts.LeaderValue()
```

Comment:
Can we just use `LeaderValue` as is and get rid of `HostName()` from the options interface? We can make `LeaderValue` default to the hostname if needed.